### PR TITLE
Sjohn/models and lora

### PR DESCRIPTION
**Pull request by sjohn4**
Here are the extra models and the Lora addapter.

### CODE REVIEW

File: modyn/config/schema/pipeline/training/config.py [hunk 1]
```diff
@@ -154,6 +160,10 @@ class TrainingConfig(ModynBaseModel):
         None,
         description="Configuration for the torch.cuda.amp.GradScaler. Effective only when amp is enabled.",
     )
+    grad_norm: float = Field(
+        default=0,
+        description="Clips the gradients normed over this value, if its 0 it will not be used.",
+    )
```
Comment (MaxiBoether) [inline modyn/config/schema/pipeline/training/config.py:166]: This is part of a different PR.

File: modyn/config/schema/pipeline/training/config.py [hunk 2]
```diff
@@ -119,6 +119,12 @@ class TrainingConfig(ModynBaseModel):
             "we start with random weights. If initial_model is 'pretrained', cannot be False."
         )
     )
+
+    lora: bool = Field(
+        False,
+        description=("Applies Lora layers to the model"),
+    )
```
Comment (MaxiBoether) [inline modyn/config/schema/pipeline/training/config.py:126]: I am a bit confused. Does this PR add only support for Lora or for multiple PEFT methods? Because if it supports more than one PEFT method, maybe we should not have a lora flag, but something like a literal PEFT field instead? I haven't properly read the source yet, but the title+description of the PR is not very clear

File: modyn/models/gpt2/gpt2.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
+from typing import Any
+
+import torch
+from torch import nn
+from transformers import AutoTokenizer, GPT2Config, GPT2LMHeadModel
+
+from modyn.models.coreset_methods_support import CoresetSupportingModule
+
+
+class Gpt2:
```
Comment (MaxiBoether) [inline modyn/models/gpt2/gpt2.py:10]: Generally, we should use `GPT2`, not `Gpt2` for consistency. Also, can we move the addition of GPT2 into a separate PR?

File: modyn/models/gpt2/gpt2.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+https://huggingface.co/docs/transformers/model_doc/gpt2
+"""
+
+
+class Gpt2Modyn(CoresetSupportingModule):
+    def __init__(self, hparams: Any) -> None:
+        super().__init__()
+
+        # Use hparams to decide the GPT-2 version
+        model_name = hparams.model_name_or_path if hasattr(hparams, "model_name_or_path") else "gpt2-large"
```
Comment (MaxiBoether) [inline modyn/models/gpt2/gpt2.py:29]: I don't get that. Isn't `model_name_or_path` super generic? Then the Gpt2 class would set this always to gpt2-large, but this reads more like a general purpose huggingface transformer class

File: modyn/models/gpt2/gpt2.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+        # Assert that the model name is valid
+        valid_model_names = {"gpt2", "gpt2-medium", "gpt2-large", "gpt2-xl"}
+        assert model_name in valid_model_names, f"Invalid model name: {model_name}. Must be one of {valid_model_names}."
+        self.tokenizer = AutoTokenizer.from_pretrained("gpt2-large")
```
Comment (MaxiBoether) [inline modyn/models/gpt2/gpt2.py:34]: use model name?

File: modyn/models/gpt2/gpt2.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+class Gpt2Modyn(CoresetSupportingModule):
+    def __init__(self, hparams: Any) -> None:
+        super().__init__()
+
+        # Use hparams to decide the GPT-2 version
+        model_name = hparams.model_name_or_path if hasattr(hparams, "model_name_or_path") else "gpt2-large"
+
+        # Assert that the model name is valid
+        valid_model_names = {"gpt2", "gpt2-medium", "gpt2-large", "gpt2-xl"}
+        assert model_name in valid_model_names, f"Invalid model name: {model_name}. Must be one of {valid_model_names}."
```
Comment (MaxiBoether) [inline modyn/models/gpt2/gpt2.py:33]: User input should be validated using exceptions. Assertions are for internal state.

File: modyn/models/gpt2/gpt2.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+        self.tokenizer.add_special_tokens(
+            {
+                "eos_token": "</s>",
+                "bos_token": "<s>",
+                "unk_token": "<unk>",
+                "pad_token": "<pad>",
+                "mask_token": "<mask>",
+            }
+        )
```
Comment (MaxiBoether) [inline modyn/models/gpt2/gpt2.py:46]: Why is this necessary`

File: modyn/models/gpt2/gpt2.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+        # Use hparams to decide the GPT-2 version
+        model_name = hparams.model_name_or_path if hasattr(hparams, "model_name_or_path") else "gpt2-large"
+
+        # Assert that the model name is valid
+        valid_model_names = {"gpt2", "gpt2-medium", "gpt2-large", "gpt2-xl"}
+        assert model_name in valid_model_names, f"Invalid model name: {model_name}. Must be one of {valid_model_names}."
+        self.tokenizer = AutoTokenizer.from_pretrained("gpt2-large")
+        # self.tokenizer.pad_token = self.tokenizer.eos_token
+        # Load the specified GPT-2 model
```
Comment (MaxiBoether) [inline modyn/models/gpt2/gpt2.py:36]: ? please clean this up a bit

File: modyn/models/gpt2/gpt2.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+          the last dimension contains token IDs and attention masks.
+        - labels (torch.Tensor, optional): Tensor of labels for language modeling tasks.
+
+        Returns:
+        - output: The output logits or loss from the GPT-2 model.
+        """
+        # Split input into token IDs and attention masks
+        input_ids = data[:, :, 0]
+
+        attention_mask = data[:, :, 1]
```
Comment (MaxiBoether) [inline modyn/models/gpt2/gpt2.py:68]: Why do we have to set the attention mask that way? not clear to me right now

File: modyn/models/gpt2/gpt2.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        """Retrieve the last layer (lm_head) of the model.
+
+        Returns:
+            The final linear layer of the GPT-2 model.
+        """
+        return self.model.lm_head
+
+    def freeze_params(self) -> None:
+        for par in self.model.parameters():
+            par.requires_grad = False
```
Comment (MaxiBoether) [inline modyn/models/gpt2/gpt2.py:85]: Do we actually use this somewhere? It's a bit weird to have this because the parent class does not require this. I feel like in terms of abstractions this is not optimal.

File: modyn/models/gpt2/gpt2.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+    #  for par in self.transformer.parameters():
+    # par.requires_grad = False
+
+    def unfreeze_params(self) -> None:
+        for par in self.model.parameters():
+            par.requires_grad = True
+        # for par in self.transformer.parameters():
+        #  par.requires_grad = True
```
Comment (MaxiBoether) [inline modyn/models/gpt2/gpt2.py:94]: clean

File: modyn/models/gpt2/gpt2.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    def generate(
+        self,
+        input_ids: torch.tensor,
+        max_length: int = 50,
+        temperature: float = 1.0,
+        top_k: int = 50,
+        top_p: float = 0.95,
+        num_return_sequences: int = 1,
+    ) -> list:
+        # Generate output sequences
```
Comment (MaxiBoether) [inline modyn/models/gpt2/gpt2.py:105]: what do we inside modyn use this for?

File: modyn/models/gpt2/others.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    AutoModelForCausalLM,
+    AutoTokenizer,
+    RobertaForMaskedLM,
+    T5ForConditionalGeneration,
+)
+
+from modyn.models.coreset_methods_support import CoresetSupportingModule
+
+
+class Roberta:
```
Comment (MaxiBoether) [inline modyn/models/gpt2/others.py:15]: why does all of this live inside gpt2/others? i think the PR might need some more cleaning/splititng

File: modyn/models/modular_adapters/modular_adapters.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
+from peft import LoraConfig, TaskType, get_peft_model
```
Comment (MaxiBoether) [inline modyn/models/modular_adapters/modular_adapters.py:1]: Please add some docstrings on how (and where) those  functions are supposed to be used

File: modyn/models/tokenizers/gpt2_tokenizer.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+        """Adapted from an example implementation of a GPT-2 tokenizer.
+
+        This implementation uses the GPT-2 tokenizer from Hugging Face's
+        Transformers library:
+        https://huggingface.co/docs/transformers/model_doc/gpt2
+        """
+        tokenizer = GPT2Tokenizer.from_pretrained("gpt2-large")
+        # tokenizer.pad_token = tokenizer.eos_token  # Set pad token to eos token to avoid padding errors
+        # tokenizer.padding_side = "right"
+        tokenizer.add_special_tokens(
```
Comment (MaxiBoether) [inline modyn/models/tokenizers/gpt2_tokenizer.py:17]: Repetition from gpt model? Also, why does the model have the tokenizer in the first place, if we also have this class?

File: modyn/tests/models/test_modular_adapters.py [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+from transformers import AutoTokenizer
+
+from modyn.models import Gpt2, apply_lora
+
+# =============================================================================
+# Updated Test Suite
+# =============================================================================
+
+# Note: The tests below assume that you're using a GPT-2 model from transformers.
+# If you want to use your custom Gpt2 class from modyn.models, adjust the imports accordingly.
```
Comment (MaxiBoether) [inline modyn/tests/models/test_modular_adapters.py:11]: When using GenAI, make sure that all of the stuff you push also is clean

File: modyn/trainer_server/custom_lr_schedulers/WarmupDecayLR/Readme.txt [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+
+This folder contains the implementation of the learning rate scheduler used by the Deepspeed library.
+We use the implementation provided by the Microsoft [repository](https://github.com/microsoft/DeepSpeed/blob/1640f6df4f6576636c472ba47ce76a507e3b4373/deepspeed/runtime/lr_schedules.py#L723)
+
+# Changes to Microsoft Version
+
+In according with the LICENSE of the NVIDIA code, we list our changes here:
+
+- We move the learning rate scheduler to a separate file.
+- We rename the class `LearningRateScheduler` to `DLRMScheduler`
```
Comment (MaxiBoether) [inline modyn/trainer_server/custom_lr_schedulers/WarmupDecayLR/Readme.txt:11]: ? dlrm? warmupdecaylr?

File: modyn/trainer_server/custom_lr_schedulers/WarmupDecayLR/_init_.py [hunk 18]
```diff
@@ -0,0 +1,7 @@
+"""DLRM LR Scheduler."""
```
Comment (MaxiBoether) [inline modyn/trainer_server/custom_lr_schedulers/WarmupDecayLR/_init_.py:1]: DLRM?

File: modyn/trainer_server/custom_lr_schedulers/WarmupDecayLR/warmupdecay.py [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+    raise TypeError(f"{type(optimizer).__name__} is not a subclass of torch.optim.Optimizer")
+
+
+def update_lr(param_groups: list[dict[str, Any]], lrs: list[float]) -> list[float]:
+    for param_group, lr in zip(param_groups, lrs):
+        param_group["lr"] = lr
+    return [group["lr"] for group in param_groups]
+
+
+class WarmupLR:
```
Comment (MaxiBoether) [inline modyn/trainer_server/custom_lr_schedulers/WarmupDecayLR/warmupdecay.py:26]: Do we really have to copy all of this here? Also should go in separate PR. Please use a library for this if we somehow can.

**Review by MaxiBoether — CHANGES_REQUESTED** [review]
Thanks for the work! I think this PR requires some more polishing at splitting

File: modyn/config/schema/pipeline/training/config.py (hunk 2 above)
Comment (sjohn4) [reply modyn/config/schema/pipeline/training/config.py:126]: Hey so this only adds Lora layers, we could have a peft field that specifies what we want to add.

File: modyn/models/gpt2/gpt2.py (hunk 12 above)
Comment (sjohn4) [reply modyn/models/gpt2/gpt2.py:105]: Hey so unfreeze, freeze parameters and  genereta are used for evaluation, generate for evaluating how it continues a asentence and freeze and unfreeze are part of the light tuning feature I built, but I could just delete them for now and put them back on the respective Prs of those features.

File: modyn/models/gpt2/gpt2.py (hunk 6 above)
Comment (sjohn4) [reply modyn/models/gpt2/gpt2.py:33]: Something like this?

File: modyn/trainer_server/custom_lr_schedulers/WarmupDecayLR/warmupdecay.py (hunk 19 above)
Comment (sjohn4) [reply modyn/trainer_server/custom_lr_schedulers/WarmupDecayLR/warmupdecay.py:26]: hey I found it as part of the deepspeed library, I will delete this and integrate this in a different Pr

File: modyn/trainer_server/custom_lr_schedulers/WarmupDecayLR/_init_.py (hunk 18 above)
Comment (sjohn4) [reply modyn/trainer_server/custom_lr_schedulers/WarmupDecayLR/_init_.py:1]: This I just copied from the other custom sheduler without even opening the function I will admit.
